Add meson build system support

TingPing opened the pull request:
This is very useful on Windows and it matches the rest of the stack such as `gobject-introspection` and `glib` which are moving to meson.

psychon commented on meson.build:
```diff
@@ -0,0 +1,34 @@
+  version: '0.9.2',
+  meson_version: '>= 0.44.0',
+  default_options: [
+    'warning_level=2',
```
According to https://github.com/mesonbuild/meson/blob/master/docs/markdown/Builtin-options.md, this should be `warnlevel`...?
Also, I have no idea what this option actually does... but whatever.

TingPing replied:
Looks like both are valid, I've never seen `warnlevel` actually used, discussing with upstream if that should be more clearly documented.

It translates to `-Wall -Wextra` with `gcc/clang`

psychon commented on meson.build:
```diff
@@ -0,0 +1,34 @@
+)
+
+lua_name = get_option('lua-pc')
+lua_dep = dependency(lua_name)
```
Would it make sense to replace this option with a brute force search for all the possible options that were seen "in the wild"?

TingPing replied:
Well, I won't maintain that list. If you will then its up to you.

psychon replied:
I think the list would just be `lua, lua5.1, lua51, lua5.2, lua52, lua5.3, lua53`. That does not seem too bad to me, but perhaps it is nicer to require people to specify this explicitly. That way it is easier to "force switch" to another Lua version.
Is it possible to customise the error message here to tell people how to specify this? E.g., how about changing the default value of `lua-pc` to "unspecified", if this is the value seen in `meson.build`, then something like `warning("You can specify the name of the Lua pkg-config file via -Dlua-pc=lua53")` is done (and perhaps then a default of `lua`) is used?
I had to Google to figure out how to set this option. Someone who tries to just run `meson` won't even know that this option exist.

Does either of that make sense?

Also: Please ping me if you think everything is addressed and I will likely just merge this without taking a second look.

CC @ILMostro just FYI and perhaps this helps simplify building on Gentoo (at the expense of depending on meson).

TingPing replied:
Sure we can do that.

TingPing commented on meson_options.txt:
```diff
@@ -0,0 +1,9 @@
+  description: 'path to lua binary to test against'
+)
+option('tests', type: 'boolean', value: true,
+  description: 'build tests'
```
Yes.

TingPing commented on lgi/meson.build:
```diff
@@ -0,0 +1,46 @@
+    dependency('gmodule-2.0'),
+  ],
+  name_prefix: '',
+  pic: true,
```
Will fix, along with other `WARNINGS`.

TingPing commented:
> The motivation is said to be Windows. Did you test this on Windows?

It has been a while but I probably did. Eitherway I ship `lgi` on Windows and will ensure it works in the long term.

TingPing commented on meson.build:
```diff
@@ -0,0 +1,34 @@
+project('lgi', 'c',
+  version: '0.9.2',
+  meson_version: '>= 0.44.0',
```
`meson` should warn now when you use too new of features (maybe not for everything as that's relatively new). I can't tell you off-hand since I wrote this months ago rather quickly. It may have been an arbitrary version.

TingPing commented on tests/meson.build:
```diff
@@ -0,0 +1,35 @@
+  env: [
+    'LD_LIBRARY_PATH=' + meson.current_build_dir(),
+    'GI_TYPELIB_PATH=' + meson.current_build_dir(),
+    'LUA_PATH=@0@/?.lua;@1@/?.lua'.format(meson.source_root(), meson.build_root()),
```
> does meson guarantee that this one is build before the test is run?)

`version.lua` uses `configure_file()` which is generated at configure time so it is guaranteed to exist before build time.

TingPing commented on tests/meson.build:
```diff
@@ -0,0 +1,35 @@
+    'GI_TYPELIB_PATH=' + meson.current_build_dir(),
+    'LUA_PATH=@0@/?.lua;@1@/?.lua'.format(meson.source_root(), meson.build_root()),
+    # FIXME: .dll on windows
+    'LUA_CPATH=@0@/?.so;'.format(meson.build_root())
```
Probably fine.

TingPing commented:
> default must be in get_pkgconfig_variable which is not really documented, except for ". (Added 0.45.0) A warning is issued if the variable is not defined, unless a default parameter is specified."
But my version is newer than that?!?

https://github.com/mesonbuild/meson/pull/3947

TingPing commented on lgi/meson.build:
```diff
@@ -0,0 +1,46 @@
+  configuration: conf,
+  install: true,
+  install_dir: join_paths(lua_path, 'lgi'),
+)
```
Done.

psychon commented:
This is likely a bad idea since it might be too "automagic" for people to figure out, but thanks to the `#mesonbuild` IRC channel, the following could be done, in theory and if we ever want this:
```
lua_dep = disabler()
foreach d : ['lua', 'lua5.1', 'lua51', 'lua5.2', 'lua52', 'lua5.3', 'lua53', 'luajit']
    if not lua_dep.found()
        lua_dep = dependency(d, required: false)
    endif
endforeach
```

TingPing commented:
I think I've addressed all the comments. I was going to add this to travis for testing but it requires a lot more work since you are still using Trusty, but if you ever move to 16.04 or newer it would be easy to test.

psychon commented:
Thanks!

> since you are still using Trusty, but if you ever move to 16.04 or newer it would be easy to test.

I'd blame Travis for only offering outdated Ubuntu versions. Yes, Docker is a thing, but it is also just a work-around.

TingPing commented:
> I'd blame Travis for only offering outdated Ubuntu versions. Yes, Docker is a thing, but it is also just a work-around.

Absolutely, If you want though I can change it to use Docker + 16.04 (or whatever).
